File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshWrappedSource.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.base.ssh;

import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.commons.util.AutoCloseableIterator;
import io.airbyte.commons.util.AutoCloseableIterators;
import io.airbyte.integrations.base.AirbyteTraceMessageUtility;
import io.airbyte.integrations.base.Source;
import io.airbyte.protocol.models.v0.AirbyteCatalog;
import io.airbyte.protocol.models.v0.AirbyteConnectionStatus;
import io.airbyte.protocol.models.v0.AirbyteConnectionStatus.Status;
import io.airbyte.protocol.models.v0.AirbyteMessage;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
import io.airbyte.protocol.models.v0.ConnectorSpecification;
import java.util.List;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SshWrappedSource implements Source {

  private static final Logger LOGGER = LoggerFactory.getLogger(SshWrappedSource.class);
  private final Source delegate;
  private final List<String> hostKey;
  private final List<String> portKey;
  private final Optional<String> sshGroup;

  public SshWrappedSource(final Source delegate, final List<String> hostKey, final List<String> portKey) {
    this.delegate = delegate;
    this.hostKey = hostKey;
    this.portKey = portKey;
    this.sshGroup = Optional.empty();
  }

  public SshWrappedSource(final Source delegate, final List<String> hostKey, final List<String> portKey, final String sshGroup) {
    this.delegate = delegate;
    this.hostKey = hostKey;
    this.portKey = portKey;
    this.sshGroup = Optional.of(sshGroup);
  }

  @Override
  public ConnectorSpecification spec() throws Exception {
    return SshHelpers.injectSshIntoSpec(delegate.spec(), sshGroup);
  }

  @Override
  public AirbyteConnectionStatus check(final JsonNode config) throws Exception {
    try {
      return SshTunnel.sshWrap(config, hostKey, portKey, delegate::check);
    } catch (final RuntimeException e) {
      final String sshErrorMessage = "Could not connect with provided SSH configuration. Error: " + e.getMessage();
      AirbyteTraceMessageUtility.emitConfigErrorTrace(e, sshErrorMessage);
      return new AirbyteConnectionStatus()
          .withStatus(Status.FAILED)
          .withMessage(sshErrorMessage);
    }
  }

  @Override
  public AirbyteCatalog discover(final JsonNode config) throws Exception {
    return SshTunnel.sshWrap(config, hostKey, portKey, delegate::discover);
  }

  @Override
  public AutoCloseableIterator<AirbyteMessage> read(final JsonNode config, final ConfiguredAirbyteCatalog catalog, final JsonNode state)
      throws Exception {
    final SshTunnel tunnel = SshTunnel.getInstance(config, hostKey, portKey);
    final AutoCloseableIterator<AirbyteMessage> delegateRead;
    try {
      delegateRead = delegate.read(tunnel.getConfigInTunnel(), catalog, state);
    } catch (final Exception e) {
      LOGGER.error("Exception occurred while getting the delegate read iterator, closing SSH tunnel", e);
      tunnel.close();
      throw e;
    }
    return AutoCloseableIterators.appendOnClose(delegateRead, tunnel::close);
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/ssh/SshWrappedSource.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.base.ssh

import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.cdk.integrations.base.AirbyteTraceMessageUtility
import io.airbyte.cdk.integrations.base.Source
import io.airbyte.commons.util.AutoCloseableIterator
import io.airbyte.commons.util.AutoCloseableIterators
import io.airbyte.protocol.models.v0.*
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.*

private val LOGGER = KotlinLogging.logger {}

class SshWrappedSource : Source {
    private val delegate: Source
    private val hostKey: List<String>
    private val portKey: List<String>
    private val sshGroup: Optional<String>

    constructor(delegate: Source, hostKey: List<String>, portKey: List<String>) {
        this.delegate = delegate
        this.hostKey = hostKey
        this.portKey = portKey
        this.sshGroup = Optional.empty()
    }

    constructor(delegate: Source, hostKey: List<String>, portKey: List<String>, sshGroup: String) {
        this.delegate = delegate
        this.hostKey = hostKey
        this.portKey = portKey
        this.sshGroup = Optional.of(sshGroup)
    }

    @Throws(Exception::class)
    override fun spec(): ConnectorSpecification {
        return SshHelpers.injectSshIntoSpec(delegate.spec(), sshGroup)
    }

    @Throws(Exception::class)
    override fun check(config: JsonNode): AirbyteConnectionStatus? {
        try {
            return SshTunnel.sshWrap(config, hostKey, portKey) { c: JsonNode -> delegate.check(c) }
        } catch (e: RuntimeException) {
            val sshErrorMessage =
                "Could not connect with provided SSH configuration. Error: " + e.message
            AirbyteTraceMessageUtility.emitConfigErrorTrace(e, sshErrorMessage)
            return AirbyteConnectionStatus()
                .withStatus(AirbyteConnectionStatus.Status.FAILED)
                .withMessage(sshErrorMessage)
        }
    }

    @Throws(Exception::class)
    override fun discover(config: JsonNode): AirbyteCatalog {
        return SshTunnel.sshWrap(config, hostKey, portKey) { c: JsonNode -> delegate.discover(c) }
    }

    @Throws(Exception::class)
    override fun read(
        config: JsonNode,
        catalog: ConfiguredAirbyteCatalog,
        state: JsonNode?
    ): AutoCloseableIterator<AirbyteMessage> {
        val tunnel: SshTunnel = SshTunnel.getInstance(config, hostKey, portKey)
        val delegateRead: AutoCloseableIterator<AirbyteMessage>
        try {
            delegateRead = delegate.read(tunnel.configInTunnel, catalog, state)
        } catch (e: Exception) {
            LOGGER.error(e) {
                "Exception occurred while getting the delegate read iterator, closing SSH tunnel"
            }
            tunnel.close()
            throw e
        }
        return AutoCloseableIterators.appendOnClose(delegateRead) { tunnel.close() }
    }

    @Throws(Exception::class)
    override fun readStreams(
        config: JsonNode,
        catalog: ConfiguredAirbyteCatalog,
        state: JsonNode?
    ): Collection<AutoCloseableIterator<AirbyteMessage>>? {
        val tunnel: SshTunnel = SshTunnel.getInstance(config, hostKey, portKey)
        try {
            return delegate.readStreams(tunnel.configInTunnel, catalog, state)
        } catch (e: Exception) {
            LOGGER.error(e) {
                "Exception occurred while getting the delegate read stream iterators, closing SSH tunnel"
            }
            tunnel.close()
            throw e
        }
    }

    companion object {}
}
